### PR TITLE
[fork] Impose a minimum transaction size of 100 bytes

### PR DESCRIPTION
**Pull request by dagurval**
https://reviews.bitcoinabc.org/D144
https://reviews.bitcoinabc.org/D1611

### CODE REVIEW

**Comment by dagurval** [comment]
@sickpig @ptschip @schancel You may find commit `[qa] Introduce bloat_tx and bloat_raw_tx` useful.

Also I've extended the ABC test to test ATMP in `[qa] Add ATMP test for undersize transaction`

**Comment by dagurval** [comment]
@xt-bot retest this

**Comment by schancel** [comment]
@dagurval That's basically a helper pulled out of the other test? txtools module is probably a good thing to start.  I've had to do a lot of fixing up on the tests that generate transactions.

**Comment by dagurval** [comment]
@schancel Sorry, I should explain. It looks to me that ABC has not attempted to activate the November upgrade on the QA test framework.

In our case when I did this, several of the qa tests started to fail due to tx-undersize. I think you will have this issue too.

I just wanted to point out here is a solution if you want it.

File: qa/rpc-tests/bip68-sequence.py [hunk 1]
```diff
@@ -235,7 +237,8 @@ def test_nonzero_locks(orig_tx, node, relayfee, use_height_lock):
             tx.nVersion = 2
             tx.vin = [CTxIn(COutPoint(orig_tx.sha256, 0), nSequence=sequence_value)]
             tx.vout = [CTxOut(int(orig_tx.vout[0].nValue - relayfee*COIN), CScript([b'a']))]
-            tx.rehash()
+            bloat_tx(tx)
+            tx.calc_sha256()
```
Comment (jasonbcox) [inline qa/rpc-tests/bip68-sequence.py]: Why switch from tx.rehash() to tx.calc_sha256()?  They're effectively equal here (and the other places where this change was made)

File: qa/rpc-tests/p2p-fullblocktest.py [hunk 2]
```diff
@@ -629,6 +632,7 @@ def update_block(block_number, new_transactions):
             scriptSig = CScript([sig, redeem_script])
 
             tx.vin[1].scriptSig = scriptSig
+            bloat_tx(tx)
```
Comment (jasonbcox) [inline qa/rpc-tests/p2p-fullblocktest.py]: Wouldn't calling bloat_tx() in tx.rehash() would make this change more pervasive and less error prone? (not to mention the nice side effect of fewer line-changes :) )

File: qa/rpc-tests/test_framework/blocktools.py [hunk 3]
```diff
@@ -74,6 +73,8 @@ def create_transaction(prevtx, n, sig, value, scriptPubKey=CScript()):
     tx.vin.append(CTxIn(COutPoint(prevtx.sha256, n), sig, 0xffffffff))
     tx.vout.append(CTxOut(value, scriptPubKey))
     tx.calc_sha256()
```
Comment (jasonbcox) [inline qa/rpc-tests/test_framework/blocktools.py]: This should come after the padding step.

File: qa/rpc-tests/p2p-fullblocktest.py (hunk 2 above)
Comment (jasonbcox) [reply qa/rpc-tests/p2p-fullblocktest.py]: Then again, having that side effect where it's not expected might make debugging difficult...

File: qa/rpc-tests/bip68-sequence.py (hunk 1 above)
Comment (dagurval) [reply qa/rpc-tests/bip68-sequence.py]: It's just a small optimization, `calc_sha256` only calculates the hash if it has not been done before. I know bloat_tx does a rehash, but I don't want to depend on this side effect.

File: qa/rpc-tests/test_framework/blocktools.py (hunk 3 above)
Comment (dagurval) [reply qa/rpc-tests/test_framework/blocktools.py]: Thanks :+1:

File: qa/rpc-tests/p2p-fullblocktest.py (hunk 2 above)
Comment (jasonbcox) [reply qa/rpc-tests/p2p-fullblocktest.py]: Ya, realized my mistake when I started poking around at the code some more myself.

File: qa/rpc-tests/test_framework/blocktools.py (hunk 3 above)
Comment (dgenr8) [reply qa/rpc-tests/test_framework/blocktools.py]: Needs fix

**Comment by dgenr8** [comment]
Rebased on top of #485.

**Pull request merged**